ouroboros-consensus/src/Ouroboros/Storage/FS/Handle.hs

```diff
@@ -0,0 +1,61 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Ouroboros.Storage.FS.Handle where
+
+import           Control.Concurrent.MVar
+import           Control.Exception
+import           System.IO.Error as IO
+
+-- import qualified Ouroboros.Storage.IO as F
+import           Ouroboros.Storage.FS.API.Types
+
+-- | File handlers for the IO instance for HasFS. 
+-- This is parametric on the os.
+--
+-- The file descriptor is provided by the HasFs and is used for Eq instance of the Handle.
+-- The 'FilePath' is used to improve error messages.
+-- The 'MVar' is used to implement 'close'.
+-- osHandle is Fd for unix and HANDLE for Windows.
+data HandleOS osHandle = Handle {
+      fid        :: Int
+    , handlePath :: FsPath
+    , filePath   :: FilePath
+    , handle     :: (MVar (Maybe osHandle))
+    }
+
+instance Show (HandleOS a) where
+  show h = "<Handle " ++ filePath h ++ ">"
+
+instance Eq (HandleOS a) where
+  h1 == h2 = fid h1 == fid h2
+
+-- | This is a no-op when the handle is already closed.
+closeHandleOS :: HandleOS osHandle -> (osHandle -> IO ())-> IO ()
+closeHandleOS (Handle _ _ _ hVar) close =
+  modifyMVar hVar $ \case
+    Nothing -> return (Nothing, ())
+    Just h -> close h >> return (Nothing, ())
+
+{-------------------------------------------------------------------------------
+  Exceptions
+-------------------------------------------------------------------------------}
+
+withOpenHandle :: String -> HandleOS osHandle -> (osHandle -> IO a) -> IO a
```

Comment:
Let's document that this is only used for the implementation of individual file system commands: the scope of the continuation is just a single command. Let's be in fact be explicit, and point out that using `withOpenHandle` for larger scopes would not be correct (since then we would not notice inside that scope if the handle would be closed).